src/app/reporting/ReportScheduler.h

```diff
@@ -0,0 +1,150 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/ReadHandler.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/IntrusiveList.h>
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace app {
+namespace reporting {
+
+// Forward declaration of TestReportScheduler to allow it to be friend with ReportScheduler
+class TestReportScheduler;
+
+using Timestamp = System::Clock::Timestamp;
+
+class ReportScheduler : public ReadHandler::Observer
+{
+public:
+    /// @brief This class acts as an interface between the report scheduler and the system timer to reduce dependencies on the
+    /// system layer.
+    class TimerDelegate
+    {
+    public:
+        virtual ~TimerDelegate() {}
+        /// @brief Start a timer for a given context, the report scheduler will always start by trying to cancel an existing timer
+        /// before starting a new one
+        /// @param context context to pass to the timer callback, in this case, a read handler node is passed
```

Comment:
```suggestion
        /// @brief Start a timer for a given context. The report scheduler must always cancel an existing timer for a context (using CancelTimer)
        /// before starting a new one for that context.
        /// @param context context to pass to the timer callback.
```
Once we do some sort of coalescing, I expect the context might not be a handler node...